File: C++/BehaviorTreesLib/Inverter.cpp
#include "Inverter.h"

namespace fluentBehaviorTree
{
	Node* Inverter::copy()
	{
		Inverter* newNode = new Inverter(this->getName());
		newNode->addChild(((this->mChild)->copy()));
		return newNode;
	}

	Inverter::Inverter(std::string name)
	{
		this->setName(name);
	}

	Inverter::~Inverter()
	{
		if (mChild != nullptr)
		{
			delete mChild;
			mChild = nullptr;
		}
	}

	// Works as NOT logic operator
	EStatus Inverter::tickNode()
	{
		this->setResult(mChild->tick());

		if (this->getResult() == EStatus::SUCCESS) this->setResult(EStatus::FAILURE);
		else if (this->getResult() == EStatus::FAILURE) this->setResult(EStatus::SUCCESS);

		return this->getResult();
	}
}

File: C#/BehaviorTree/Branch/Decorator/Inverter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FluentBehaviorTree
{
    public class Inverter : Decorator
    {
        public Inverter(string name)
        {
            this.Name = name;
        }

        public override Node Copy()
        {
            Inverter newNode = new Inverter(this.Name);
            newNode.AddChild(this.child.Copy());
            return newNode;
        }

        /// <summary>
        /// Works as NOT logic operator
        /// </summary>
        /// <returns></returns>
        protected override Status tick()
        {
            this.Result = child.Tick();

            if (this.Result == Status.SUCCESS) this.Result = Status.FAILURE;
            else if (this.Result == Status.FAILURE) this.Result = Status.SUCCESS;

            return this.Result;
        }
    }
}
